jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/DefaultServletCombinationsTest.java

```diff
@@ -0,0 +1,191 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+@ExtendWith(WorkDirExtension.class)
+public class DefaultServletCombinationsTest
+{
+    public WorkDir workDir;
+
+    public Path docRoot;
+    private Server server;
+    private LocalConnector connector;
+    private ServletContextHandler context;
+
+    @BeforeEach
+    public void init() throws Exception
+    {
+        docRoot = workDir.getEmptyPathDir().resolve("docroot");
+        FS.ensureDirExists(docRoot);
+        Path subdirPath = docRoot.resolve("subdir");
+        FS.ensureDirExists(subdirPath);
+        Files.writeString(docRoot.resolve("index.html"), "Static index.html at root", UTF_8);
+        Files.writeString(docRoot.resolve("foo.welcome"), "Static foo.welcome at root", UTF_8);
+        Files.writeString(subdirPath.resolve("index.html"), "Static index.html at subdir", UTF_8);
+        Files.writeString(subdirPath.resolve("foo.welcome"), "Static foo.welcome at subdir", UTF_8);
+
+        server = new Server();
+
+        connector = new LocalConnector(server);
+        connector.getConnectionFactory(HttpConfiguration.ConnectionFactory.class).getHttpConfiguration().setSendServerVersion(false);
+
+        context = new ServletContextHandler();
+        context.setBaseResourceAsPath(docRoot);
+        context.setContextPath("/");
+        context.setWelcomeFiles(new String[]{"index.html", "index.welcome"});
+
+        ServletHolder welcomeExtHolder = context.addServlet(WelcomeServlet.class, "*.welcome");
+        welcomeExtHolder.setInitParameter("mapping", "welcome extension");
+    }
+
+    private void startServer(boolean pathInfoOnly, ResourceService.WelcomeMode welcomeMode) throws Exception
+    {
+        ServletHolder defaultHolder = context.addServlet(DefaultServlet.class, "/");
+        defaultHolder.setInitParameter("pathInfoOnly", String.valueOf(pathInfoOnly));
+
+        server.setHandler(context);
+        server.addConnector(connector);
+
+        server.start();
+        ((DefaultServlet)defaultHolder.getServlet()).getResourceService().setWelcomeMode(welcomeMode);
+    }
+
+    @AfterEach
+    public void destroy() throws Exception
+    {
+        server.stop();
+        server.join();
+    }
+
+    public static class WelcomeServlet extends HttpServlet
+    {
+        @Override
+        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException
+        {
+            resp.setStatus(200);
+            String mapping = getInitParameter("mapping");
+            resp.getWriter().print("Servlet at " + mapping);
+        }
+    }
+
+    record Data(boolean pathInfoOnly, ResourceService.WelcomeMode welcomeMode, String requestPath, int expectedStatus, String expected)
+    {
+    }
+
+    public static Stream<Data> data()
+    {
+        return Stream.of(
+            new Data(false, SERVE, "/", HttpStatus.OK_200, "Static index.html at root"),
```

Comment:
I don't like flattened lists like this as it is impossible to tell if all combinations have been handled, I'd much rather see:
```javav
    for(boolean pathInfoOnly : new boolean[]{ true, false})
        for (WelcomeMode m : new WelcomeMode[]{REDIRECT, REHANDLE, SERVE})
           for (String uri : .... etc.)
```

I know this means you can't list out the expected results, but you can write a little expert system to produce the expected result... i.e a few if elses 

```